x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportEvaluateDataFrameAction.java

```diff
@@ -38,24 +39,64 @@ public TransportEvaluateDataFrameAction(TransportService transportService, Actio
     @Override
     protected void doExecute(Task task, EvaluateDataFrameAction.Request request,
                              ActionListener<EvaluateDataFrameAction.Response> listener) {
-        Evaluation evaluation = request.getEvaluation();
-        SearchRequest searchRequest = new SearchRequest(request.getIndices());
-        searchRequest.source(evaluation.buildSearch(request.getParsedQuery()));
-
-        ActionListener<List<EvaluationMetricResult>> resultsListener = ActionListener.wrap(
-            results -> listener.onResponse(new EvaluateDataFrameAction.Response(evaluation.getName(), results)),
+        ActionListener<List<Void>> resultsListener = ActionListener.wrap(
+            unused -> {
+                EvaluateDataFrameAction.Response response =
+                    new EvaluateDataFrameAction.Response(request.getEvaluation().getName(), request.getEvaluation().getResults());
+                listener.onResponse(response);
+            },
             listener::onFailure
         );
 
-        client.execute(SearchAction.INSTANCE, searchRequest, ActionListener.wrap(
-            searchResponse -> threadPool.generic().execute(() -> {
-                try {
-                    evaluation.evaluate(searchResponse, resultsListener);
-                } catch (Exception e) {
-                    listener.onFailure(e);
-                };
-            }),
-            listener::onFailure
-        ));
+        EvaluationExecutor evaluationExecutor = new EvaluationExecutor(threadPool, client, request);
+        // Add one task only. Other tasks will be added as needed during execution.
+        evaluationExecutor.add(evaluationExecutor.newTask());
```

Comment:
Perhaps we should add the first task in the constructor of the executor and then we won't need this at all.

Reply:
Done.